File: apps/web/src/lib/api/grants.ts
import { apiRequest, apiRequestWithMeta } from "@/lib/api/http";
import type {
  AccessGrantResponse,
  GrantInvitationResponse,
  GrantListResponse,
  GrantRevokeResponse,
} from "@/lib/api/generated/types.gen";

export async function listGrants(sessionToken: string): Promise<GrantListResponse> {
  return apiRequest<GrantListResponse>("/v1/grants", { sessionToken });
}

export async function createGrantInvitation(
  sessionToken: string,
  body: {
    object_kind: "component" | "setup";
    stable_id: string;
    major: number;
    recipient_email: string;
    idempotency_key: string;
  },
): Promise<{ body: GrantInvitationResponse; operationId: string | null }> {
  const result = await apiRequestWithMeta<GrantInvitationResponse>("/v1/grants/invitations", {
    method: "POST",
    sessionToken,
    body: {
      schema_version: 1,
      ...body,
      ttl_seconds: 604_800,
    },
  });
  return { body: result.data, operationId: result.operationId };
}

export async function createDirectGrant(
  sessionToken: string,
  body: {
    object_kind: "component" | "setup";
    stable_id: string;
    major: number;
    recipient_kind: "github_username" | "user_id";
    recipient: string;
    idempotency_key: string;
  },
): Promise<{ body: AccessGrantResponse; operationId: string | null }> {
  const result = await apiRequestWithMeta<AccessGrantResponse>("/v1/grants/direct", {
    method: "POST",
    sessionToken,
    body: { schema_version: 1, ...body },
  });
  return { body: result.data, operationId: result.operationId };
}

export async function acceptGrantInvitation(
  sessionToken: string,
  invitationId: string,
  token: string,
  idempotencyKey: string,
): Promise<{ body: AccessGrantResponse; operationId: string | null }> {
  const result = await apiRequestWithMeta<AccessGrantResponse>(
    `/v1/grants/invitations/${invitationId}/accept`,
    {
      method: "POST",
      sessionToken,
      body: {
        schema_version: 1,
        token,
        idempotency_key: idempotencyKey,
      },
    },
  );
  return { body: result.data, operationId: result.operationId };
}

export async function revokeGrantInvitation(
  sessionToken: string,
  invitationId: string,
  reason: string,
  idempotencyKey: string,
): Promise<{ body: GrantRevokeResponse; operationId: string | null }> {
  const result = await apiRequestWithMeta<GrantRevokeResponse>(
    `/v1/grants/invitations/${invitationId}/revoke`,
    {
      method: "POST",
      sessionToken,
      body: {
        schema_version: 1,
        reason,
        idempotency_key: idempotencyKey,
      },
    },
  );
  return { body: result.data, operationId: result.operationId };
}

export async function revokeAccessGrant(
  sessionToken: string,
  grantId: string,
  reason: string,
  idempotencyKey: string,
): Promise<{ body: GrantRevokeResponse; operationId: string | null }> {
  const result = await apiRequestWithMeta<GrantRevokeResponse>(`/v1/grants/${grantId}/revoke`, {
    method: "POST",
    sessionToken,
    body: {
      schema_version: 1,
      reason,
      idempotency_key: idempotencyKey,
    },
  });
  return { body: result.data, operationId: result.operationId };
}

File: packages/contracts/src/ai_stp_contracts/grants.py
"""Access grant and invitation wire payloads (SPEC-026, SPEC-002, ADR-0020/0030)."""

from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

from ai_stp_contracts.auth import AccountId
from ai_stp_contracts.http import (
    IdempotencyKey,
    Timestamp,
    open_wire_object,
    strict_request_object,
)

type InvitationState = Literal["pending", "accepted", "expired", "revoked"]
type GrantState = Literal["active", "revoked"]
type ObjectKind = Literal["component", "setup"]
type DirectRecipientKind = Literal["github_username", "user_id"]


class DirectGrantCreateRequest(BaseModel):
    """POST /v1/grants/direct body with an explicit recipient identifier kind."""

    model_config = ConfigDict(extra="forbid", frozen=True, json_schema_extra=strict_request_object)

    schema_version: Literal[1] = 1
    object_kind: ObjectKind
    stable_id: Annotated[str, Field(min_length=8, max_length=64)]
    major: Annotated[int, Field(ge=0, le=9999)]
    recipient_kind: DirectRecipientKind
    recipient: Annotated[str, Field(min_length=1, max_length=320)]
    idempotency_key: IdempotencyKey


class GrantInvitationCreateRequest(BaseModel):
    """POST /v1/grants/invitations body."""

    model_config = ConfigDict(extra="forbid", frozen=True, json_schema_extra=strict_request_object)

    schema_version: Literal[1] = 1
    object_kind: ObjectKind
    stable_id: Annotated[str, Field(min_length=8, max_length=64)]
    major: Annotated[int, Field(ge=0, le=9999)]
    recipient_email: Annotated[str, Field(min_length=3, max_length=320)]
    idempotency_key: IdempotencyKey
    ttl_seconds: Annotated[int, Field(default=604_800, ge=60, le=2_592_000)] = 604_800


class GrantInvitationResponse(BaseModel):
    """Invitation create/list item. Never includes the raw token."""

    model_config = ConfigDict(extra="allow", frozen=True, json_schema_extra=open_wire_object)

    schema_version: Literal[1] = 1
    invitation_id: Annotated[str, Field(min_length=8, max_length=64)]
    object_kind: ObjectKind
    stable_id: str
    major: int
    state: InvitationState
    expires_at: Timestamp
    created_at: Timestamp


class AccessGrantResponse(BaseModel):
    """One major-line access grant."""

    model_config = ConfigDict(extra="allow", frozen=True, json_schema_extra=open_wire_object)

    schema_version: Literal[1] = 1
    grant_id: Annotated[str, Field(min_length=8, max_length=64)]
    object_kind: ObjectKind
    stable_id: str
    major: int
    grantee_account_id: AccountId
    owner_account_id: AccountId
    state: GrantState
    created_at: Timestamp
    revoked_at: Timestamp | None = None
    recipient_kind: DirectRecipientKind | None
    recipient: str | None


class GrantListResponse(BaseModel):
    """Owned invitations and grants for the caller."""

    model_config = ConfigDict(extra="allow", frozen=True, json_schema_extra=open_wire_object)

    schema_version: Literal[1] = 1
    invitations: Annotated[list[GrantInvitationResponse], Field(default_factory=list)]
    grants: Annotated[list[AccessGrantResponse], Field(default_factory=list)]


class GrantAcceptRequest(BaseModel):
    """POST /v1/grants/invitations/{invitation_id}/accept body."""

    model_config = ConfigDict(extra="forbid", frozen=True, json_schema_extra=strict_request_object)

    schema_version: Literal[1] = 1
    token: Annotated[str, Field(min_length=16, max_length=256)]
    idempotency_key: IdempotencyKey


class GrantRevokeRequest(BaseModel):
    """Revoke invitation or grant with an optional reason."""

    model_config = ConfigDict(extra="forbid", frozen=True, json_schema_extra=strict_request_object)

    schema_version: Literal[1] = 1
    reason: Annotated[str, Field(default="", max_length=500)] = ""
    idempotency_key: IdempotencyKey


class GrantRevokeResponse(BaseModel):
    """Outcome of a revoke action."""

    model_config = ConfigDict(extra="allow", frozen=True, json_schema_extra=open_wire_object)

    schema_version: Literal[1] = 1
    revoked: bool = True
    local_bytes_retained: bool = True


class CliGrantInvitationView(GrantInvitationResponse):
    """CLI view of one invitation, separate from the HTTP schema boundary."""


class CliGrantAccessView(AccessGrantResponse):
    """CLI view of one access grant, separate from the HTTP schema boundary."""


class CliGrantListView(GrantListResponse):
    """CLI view of the caller's grants, separate from the HTTP schema boundary."""


class CliGrantRevokeView(GrantRevokeResponse):
    """CLI view of a revocation, separate from the HTTP schema boundary."""
